File: app/src/androidTest/java/email/schaal/ocreader/LoginActivityTest.java
package email.schaal.ocreader;

import android.support.test.filters.LargeTest;
import android.support.test.rule.ActivityTestRule;
import android.support.test.runner.AndroidJUnit4;

import org.junit.After;
import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;
import org.junit.runner.RunWith;

import java.io.IOException;

import email.schaal.ocreader.database.Queries;
import okhttp3.HttpUrl;
import okhttp3.mockwebserver.MockWebServer;

import static android.support.test.espresso.Espresso.onView;
import static android.support.test.espresso.action.ViewActions.clearText;
import static android.support.test.espresso.action.ViewActions.click;
import static android.support.test.espresso.action.ViewActions.typeText;
import static android.support.test.espresso.assertion.ViewAssertions.matches;
import static android.support.test.espresso.matcher.ViewMatchers.hasErrorText;
import static android.support.test.espresso.matcher.ViewMatchers.withId;

/**
 * Created by daniel on 13.10.16.
 */
@RunWith(AndroidJUnit4.class)
@LargeTest
public class LoginActivityTest {
    @Rule
    public ActivityTestRule<LoginActivity> activityTestRule = new ActivityTestRule<>(LoginActivity.class);

    private final MockWebServer server = new MockWebServer();

    public LoginActivityTest() {
        server.setDispatcher(new APIDispatcher());
    }

    @Before
    public void setUp() throws Exception {
        Queries.resetDatabase();
        server.start();
    }

    @After
    public void tearDown() throws Exception {
        Queries.resetDatabase();
        server.shutdown();
    }

    @Test
    public void testInsecureLogin() throws IOException {
        HttpUrl baseUrl = server.url("");

        onView(withId(R.id.url)).perform(clearText(), typeText(baseUrl.toString()));
        onView(withId(R.id.username)).perform(clearText(), typeText("admin"));
        onView(withId(R.id.password)).perform(clearText(), typeText("admin"));
        onView(withId(R.id.sign_in_button)).perform(click());
        onView(withId(R.id.url)).check(matches(hasErrorText(activityTestRule.getActivity().getString(R.string.error_insecure_connection))));
        onView(withId(R.id.sign_in_button)).perform(click());
    }

    @Test
    public void testUnknownHost() throws IOException {
        onView(withId(R.id.url)).perform(clearText(), typeText("https://unknown-host"));
        onView(withId(R.id.username)).perform(clearText(), typeText("admin"));
        onView(withId(R.id.password)).perform(clearText(), typeText("admin"));
        onView(withId(R.id.sign_in_button)).perform(click());
        onView(withId(R.id.url)).check(matches(hasErrorText(activityTestRule.getActivity().getString(R.string.error_unknown_host))));
    }
}

File: app/src/androidTest/java/email/schaal/ocreader/LoginActivityTest.kt
package email.schaal.ocreader

import androidx.test.espresso.Espresso
import androidx.test.espresso.action.ViewActions
import androidx.test.espresso.assertion.ViewAssertions
import androidx.test.espresso.matcher.ViewMatchers
import androidx.test.ext.junit.runners.AndroidJUnit4
import androidx.test.filters.LargeTest
import androidx.test.rule.ActivityTestRule
import email.schaal.ocreader.LoginActivity
import email.schaal.ocreader.api.API
import email.schaal.ocreader.database.Queries
import okhttp3.mockwebserver.MockWebServer
import org.junit.After
import org.junit.Before
import org.junit.Rule
import org.junit.Test
import org.junit.runner.RunWith
import java.io.IOException

/**
 * Created by daniel on 13.10.16.
 */
@RunWith(AndroidJUnit4::class)
@LargeTest
class LoginActivityTest {
    @Rule
    val activityTestRule = ActivityTestRule(LoginActivity::class.java)
    private val server = MockWebServer()
    private val dispatcher = APIDispatcher()
    @Before
    @Throws(Exception::class)
    fun setUp() {
        Queries.resetDatabase()
        server.start()
    }

    @After
    @Throws(Exception::class)
    fun tearDown() {
        Queries.resetDatabase()
        server.shutdown()
    }

    @Test
    @Throws(IOException::class)
    fun testInsecureLogin() {
        val baseUrl = server.url("")
        Espresso.onView(ViewMatchers.withId(R.id.url)).perform(ViewActions.clearText(), ViewActions.typeText(baseUrl.toString()))
        Espresso.onView(ViewMatchers.withId(R.id.username)).perform(ViewActions.clearText(), ViewActions.typeText("admin"))
        Espresso.onView(ViewMatchers.withId(R.id.password)).perform(ViewActions.clearText(), ViewActions.typeText("admin"))
        Espresso.onView(ViewMatchers.withId(R.id.sign_in_button)).perform(ViewActions.scrollTo(), ViewActions.click())
        Espresso.onView(ViewMatchers.withId(R.id.url)).check(ViewAssertions.matches(ViewMatchers.hasErrorText(activityTestRule.activity.getString(R.string.error_insecure_connection))))
        Espresso.onView(ViewMatchers.withId(R.id.sign_in_button)).perform(ViewActions.scrollTo(), ViewActions.click())
    }

    @Test
    @Throws(IOException::class)
    fun testUnknownHost() {
        Espresso.onView(ViewMatchers.withId(R.id.url)).perform(ViewActions.clearText(), ViewActions.typeText("https://unknown-host"))
        Espresso.onView(ViewMatchers.withId(R.id.username)).perform(ViewActions.clearText(), ViewActions.typeText("admin"))
        Espresso.onView(ViewMatchers.withId(R.id.password)).perform(ViewActions.clearText(), ViewActions.typeText("admin"))
        Espresso.onView(ViewMatchers.withId(R.id.sign_in_button)).perform(ViewActions.scrollTo(), ViewActions.click())
        Espresso.onView(ViewMatchers.withId(R.id.url)).check(ViewAssertions.matches(ViewMatchers.hasErrorText(activityTestRule.activity.getString(R.string.error_unknown_host))))
    }

    @Test
    @Throws(IOException::class)
    fun testOutdatedVersion() {
        val originalVersion = dispatcher.version
        dispatcher.version = "8.8.0"
        val baseUrl = server.url("")
        Espresso.onView(ViewMatchers.withId(R.id.url)).perform(ViewActions.clearText(), ViewActions.typeText(baseUrl.toString()))
        Espresso.onView(ViewMatchers.withId(R.id.username)).perform(ViewActions.clearText(), ViewActions.typeText("admin"))
        Espresso.onView(ViewMatchers.withId(R.id.password)).perform(ViewActions.clearText(), ViewActions.typeText("admin"))
        Espresso.onView(ViewMatchers.withId(R.id.sign_in_button)).perform(ViewActions.scrollTo(), ViewActions.click())
        Espresso.onView(ViewMatchers.withId(R.id.url)).check(ViewAssertions.matches(ViewMatchers.hasErrorText(activityTestRule.activity.getString(R.string.error_insecure_connection))))
        Espresso.onView(ViewMatchers.withId(R.id.sign_in_button)).perform(ViewActions.scrollTo(), ViewActions.click())
        Espresso.onView(ViewMatchers.withId(R.id.status)).check(ViewAssertions.matches(ViewMatchers.withText(activityTestRule.activity.getString(R.string.ncnews_too_old, API.MIN_VERSION.toString()))))
        dispatcher.version = originalVersion
    }

    init {
        server.dispatcher = dispatcher
    }
}